File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        
        Deque<Integer> belt = new ArrayDeque<>();
        Deque<Integer> temp = new ArrayDeque<>();
        
        for(int i=1; i<= order.length; i++){
            belt.add(i);
        }
        
        int cnt = 0;
        
        while(cnt < order.length){
            
            while(!belt.isEmpty() && order[cnt] > belt.peek()){
                temp.add(belt.poll());
            }
            
            if(!belt.isEmpty() && order[cnt] == belt.peek()){
                cnt++;
                belt.poll();
            }else if(!temp.isEmpty() && order[cnt] == temp. peekLast()){
                cnt++;
                temp.pollLast();
            }else break;
            
            
        }
        
        return cnt;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.js
function solution(order) {
    
    const belt = [];
    const temp = [];
    
    for(let i=order.length; i>0; i--){
        belt.push(i);
    }
        
    var index = 0;
    
    while(index < order.length){
        
        while(belt.length && order[index] > belt[belt.length-1]){
            temp.push(belt[belt.length-1]);
            belt.pop();
        }

        if(belt.length && belt[belt.length-1] == order[index]){
            index++;
            belt.pop();
        }else if(temp.length && temp[temp.length -1] == order[index]){
            index++;
            temp.pop();
        }else break;
        
    }
    
    return index;
}